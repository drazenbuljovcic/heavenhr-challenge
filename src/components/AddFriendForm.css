:local(.addFriendInput) {
  border-radius: 0;
  border-color: #ABAAAA;
  border-left: 0;
  border-right: 0;
}

:local(.friendForm) {
  display: flex;
}

:local(.genderSelectionContainer) {
  display: flex;
}

input[type=radio] {
  width: 0;
  height: 0;
}

label {
  margin: 0;
  display: flex;
  align-items: center;
}

label > i {
  font-size: 1.4em !important;
}

input[type=radio]:checked + label > i {
  color: red;
}