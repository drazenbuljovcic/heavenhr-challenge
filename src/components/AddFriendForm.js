import React, { Component } from 'react';
import { PropTypes } from 'prop-types';

import classnames from 'classnames';
import styles from './AddFriendForm.css';

class AddFriendForm extends Component {
  //#region MARK: Super methods
  constructor (props, context) {
    super(props, context);

    this.state = {
      name: '',
      gender: '',
    };

    this.handleNameChange = this.handleNameChange.bind(this);
    this.handleGenderChange = this.handleGenderChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }
  //#endregion

  //#region MARK: Event handlers
  handleNameChange(e) {
    this.setState({ name: e.target.value });
  }
  
  handleGenderChange(e) {
    this.setState({ 'gender': e.target.value });
  }

  handleSubmit (e) {
    e.preventDefault();
    
    if(!this.state.name) { return; }

    this.props.addFriend(this.state.name, this.state.gender);

    e.target.reset();
    this.input.focus();
    this.setState({ name: '', gender: '' });
  }
  //#endregion

  render () {
    return (
      <form onSubmit={this.handleSubmit} className={classnames(styles.friendForm)}>
        <input
          type="text"
          autoFocus="true"
          ref={input => this.input = input}
          className={classnames('form-control', styles.addFriendInput)}
          placeholder="Name of your friend"
          value={this.state.name}
          onChange={this.handleNameChange} />

        <div className={classnames(styles.genderSelectionContainer)}>
          <input type="radio" id="gender-male" value="male" name="gender" onChange={this.handleGenderChange} />
          <label htmlFor="gender-male">
            <i className={classnames('btn btn-default fa fa-male')} aria-hidden="true"></i>
          </label>

          <input type="radio" id="gender-female" value="female" name="gender" onChange={this.handleGenderChange} />
          <label htmlFor="gender-female">
            <i className={classnames('btn btn-default fa fa-female')} aria-hidden="true"></i>
          </label>
        </div>

        <button className={classnames('btn btn-success')} type="submit">
          <i className="fa fa-check" aria-hidden="true"></i>
        </button>
      </form>
    );
  }
}

AddFriendForm.propTypes = {
  addFriend: PropTypes.func.isRequired,
};

export default AddFriendForm;
