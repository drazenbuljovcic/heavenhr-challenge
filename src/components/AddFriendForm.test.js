import React from 'react';

import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-15';

import AddFriendForm from './AddFriendForm';

configure({ adapter: new Adapter() });

describe('Form for adding new friends', () => {
  const defaultProps = {
    addFriend: jest.fn(),
  };
  const component = shallow(<AddFriendForm {...defaultProps} />);
  const form = component.find('form');

  it('should match it\'s empty snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should have one form element', () => {
    expect(form).toHaveLength(1);
  });

  it('should have one text input element for name inside the form', () => {
    expect(form.find('input[name=\'name\']')).toHaveLength(1);
  });

  it('should have two radio inputs for gender inside the form', () => {
    expect(form.find('input[name=\'gender\']')).toHaveLength(2);
  });

  it('should have male radio button label with icon', () => {
    const maleInput = form.find('input[value=\'male\']');
    const label = form.find('input[value=\'male\'] + label');

    expect(maleInput).toHaveLength(1);
    expect(label).toHaveLength(1);
    expect(label.find('i.fa-male')).toHaveLength(1);
  });

  
  it('should have female radio button label with icon', () => {
    const femaleInput = form.find('input[value=\'female\']');
    const label = form.find('input[value=\'female\'] + label');

    expect(femaleInput).toHaveLength(1);
    expect(label).toHaveLength(1);
    expect(label.find('i.fa-female')).toHaveLength(1);
  });

  describe('Add friend', () => {
    const mockEvent = { preventDefault: () => {}, target: { reset: jest.fn() }};
    component.instance().input = { focus: jest.fn() };

    it('should add the friend if the form is valid and submitted', () => {
      Promise.resolve(component.setState({ 'name': 'Friend' }))
        .then(() => {
          component.instance().handleSubmit(mockEvent);
          expect(defaultProps.addFriend).toHaveBeenCalled();
        }).catch(() => {});
    });

    it('should not add the friend name does not exist', () => {
      Promise.resolve(component.setState({ 'name': '' }))
        .then(() => {
          component.instance().handleSubmit(mockEvent);
          expect(defaultProps.addFriend).not.toHaveBeenCalled();
        }).catch(() => {});
    });
  });
});