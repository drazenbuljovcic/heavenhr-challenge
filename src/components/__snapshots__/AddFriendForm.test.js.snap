exports[`Form for adding new friends should match it's empty snapshot 1`] = `
<form
  className=""
  onSubmit={[Function]}>
  <input
    autoFocus="true"
    className="form-control"
    name="name"
    onChange={[Function]}
    placeholder="Name of your friend"
    type="text"
    value="" />
  <div
    className="">
    <input
      id="gender-male"
      name="gender"
      onChange={[Function]}
      type="radio"
      value="male" />
    <label
      htmlFor="gender-male">
      <i
        aria-hidden="true"
        className="btn btn-default fa fa-male" />
    </label>
    <input
      id="gender-female"
      name="gender"
      onChange={[Function]}
      type="radio"
      value="female" />
    <label
      htmlFor="gender-female">
      <i
        aria-hidden="true"
        className="btn btn-default fa fa-female" />
    </label>
  </div>
  <button
    className="btn btn-success"
    type="submit">
    <i
      aria-hidden="true"
      className="fa fa-check" />
  </button>
</form>
`;
